Update api urls

ekcorso opened the pull request:
This updates the baseURL and the paths for getUserByUUID(), usePowerAtOneTimeGateway(), usePowerAtOngoingGateway(),requestTransfer(), approveTransfer().

ekcorso committed:
Update baseURL

ekcorso committed:
Update url paths

This updates getUserByUUID(), usePowerAtOneTimeGateway(),
usePowerAtOngoingGateway(),requestTransfer(), approveTransfer().

ekcorso committed:
Update url paths

This updates getUserByUUID(), usePowerAtOneTimeGateway(),
usePowerAtOngoingGateway(),requestTransfer(), approveTransfer().

zach-freyja commented on PlanetNineGateway/Common/Utilities/Network.swift:
```diff
@@ -122,18 +122,22 @@ class Network: NSObject {
     func getUserByUUID(userUUID: String, gatewayName: String, timestamp: String, signature: String, callback: @escaping (Error?, Data?) -> Void) {
-        let path = "/user/userUUID/\(userUUID)/gateway/\(gatewayName)/signature/\(signature)/timestamp/\(timestamp)"
+        //let path = "/user/userUUID/\(userUUID)/gateway/\(gatewayName)/signature/\(signature)/timestamp/\(timestamp)"
+        let path = "/user/\(gatewayName)/userUUID/\(userUUID)/signature/\(signature)/timestamp/\(timestamp)"
```
This one should be "/gateway/\(gatewayName)/userUUID/\(userUUID)/signature/\(signature)/timestamp/\(timestamp)"

zach-freyja commented on PlanetNineGateway/Common/Utilities/Network.swift:
```diff
@@ -122,18 +122,22 @@ class Network: NSObject {
     func usePowerAtOneTimeGateway(powerUsageObject: PowerUsage, callback: @escaping (Error?, Data?) -> Void) {
-        let path = "/user/userUUID/\(powerUsageObject.userUUID)/power/gateway/\(powerUsageObject.gatewayName)"
+//        let path = "/user/userUUID/\(powerUsageObject.userUUID)/power/gateway/\(powerUsageObject.gatewayName)"
+        let path = "/user/\(powerUsageObject.gatewayName)/power"
```
This one should be "/gateway/\(powerUsageObject.gatewayName)/power"

zach-freyja commented on PlanetNineGateway/Common/Utilities/Network.swift:
```diff
@@ -122,18 +122,22 @@ class Network: NSObject {
     func usePowerAtOngoingGateway(usePowerAtOngoingGatewayWithSignature: UsePowerAtOngoingGatewayWithSignature, callback: @escaping (Error?, Data?) -> Void) {
-        let path = "/user/userUUID/\(usePowerAtOngoingGatewayWithSignature.userUUID)/power/gateway/\(usePowerAtOngoingGatewayWithSignature.gatewayName)/ongoing"
+//        let path = "/user/userUUID/\(usePowerAtOngoingGatewayWithSignature.userUUID)/power/gateway/\(usePowerAtOngoingGatewayWithSignature.gatewayName)/ongoing"
+        let path = "/user/\(usePowerAtOngoingGatewayWithSignature.gatewayName)/ongoing/power"
```
again user should be gateway in this one

zach-freyja commented on PlanetNineGateway/Common/Utilities/Network.swift:
```diff
@@ -144,7 +148,8 @@ class Network: NSObject {
     func requestTransfer(transferRequestWithSignature: TransferRequestWithSignature, gatewayName: String, callback: @escaping (Error?, Data?) -> Void) {
-        let path = "/user/userUUID/\(transferRequestWithSignature.userUUID)/gateway/\(gatewayName)/transfer/request"
+//        let path = "/user/userUUID/\(transferRequestWithSignature.userUUID)/gateway/\(gatewayName)/transfer/request"
+        let path = "/user/gateway/request"
```
this should be "/transfer/gateway/request"

zach-freyja commented on PlanetNineGateway/Common/Utilities/Network.swift:
```diff
@@ -187,7 +192,8 @@ class Network: NSObject {
         ///user/userId/:userId/transfer/nineumTransactionId/:nineumTransactionId/approve'
-        let path = "/user/userId/\(approveTransferWithSignature.userId)/transfer/nineumTransactionId/\(approveTransferWithSignature.nineumTransactionId)/approve"
+//        let path = "/user/userId/\(approveTransferWithSignature.userId)/transfer/nineumTransactionId/\(approveTransferWithSignature.nineumTransactionId)/approve"
+        let path = "/user/approve"
```
This should be "/transfer/approve".

zach-freyja reviewed (changes requested):
So all these paths need to have `user` changed to either `gateway` or `transfer`. In the api map doc (https://github.com/planet-nine-app/pn-cdk/blob/main/docs/apimap.md) the root path is the header for that section. So all the gateway endpoints start with `/gateway`. I'm curious if you tested these. I'm assuming you did, and that they worked, so I'm wondering how these changes didn't get picked up since these paths would have all been broken.

Commit:
fixes

ekcorso committed:
Fix url path roots

ekcorso committed:
Add Crypto.swift file

ekcorso committed:
Add getUser and supporting methods to the gateway

Adds a getUser method to PlanetNineGateway so as to remove crpyto references from the app itself. Also adds various save methods to UserModel
and Utils, and removes commented out getUserByUUID path from Network.

ekcorso committed:
Update toSting() for UserPowerAtOngoingGateway

Removes user from string

ekcorso committed:
Add userPowerAtOngoingGateway method to gateway

Moves crypto operations here.

ekcorso committed:
Remove commented out path

ekcorso committed:
Merge Network updates

zach-freyja commented on PlanetNineGateway/iOS/Models/Network.swift:
```diff
@@ -122,22 +123,34 @@ class Network: NSObject {
+=======
         //let path = "/user/userUUID/\(userUUID)/gateway/\(gatewayName)/signature/\(signature)/timestamp/\(timestamp)"
         let path = "/user/\(gatewayName)/userUUID/\(userUUID)/signature/\(signature)/timestamp/\(timestamp)"
+>>>>>>> 54d724ec0e6d892a1ec048e5d0321dc66199a470:PlanetNineGateway/Common/Utilities/Network.swift
```
Looks like your PR is conflicting here. You can resolve these pretty easily by removing the bottom and keeping the top it looks like.

zach-freyja commented on PlanetNineGateway/iOS/Models/Network.swift:
```diff
@@ -122,22 +123,34 @@ class Network: NSObject {
+=======
 //        let path = "/user/userUUID/\(usePowerAtOngoingGatewayWithSignature.userUUID)/power/gateway/\(usePowerAtOngoingGatewayWithSignature.gatewayName)/ongoing"
         let path = "/user/\(usePowerAtOngoingGatewayWithSignature.gatewayName)/ongoing/power"
+>>>>>>> 54d724ec0e6d892a1ec048e5d0321dc66199a470:PlanetNineGateway/Common/Utilities/Network.swift
```
Same with these conflicts

zach-freyja commented on PlanetNineGateway/iOS/Models/Network.swift:
```diff
@@ -149,7 +162,11 @@ class Network: NSObject {
+=======
         let path = "/user/gateway/request"
+>>>>>>> 54d724ec0e6d892a1ec048e5d0321dc66199a470:PlanetNineGateway/Common/Utilities/Network.swift
         let jsonData = Utils().encodableToJSONData(transferRequestWithSignature)
```
And these conflicts

zach-freyja commented on PlanetNineGateway/iOS/Models/Network.swift:
```diff
@@ -193,7 +210,11 @@ class Network: NSObject {
+=======
         let path = "/user/approve"
+>>>>>>> 54d724ec0e6d892a1ec048e5d0321dc66199a470:PlanetNineGateway/Common/Utilities/Network.swift
         let jsonData = Utils().encodableToJSONData(approveTransferWithSignature)
```
And these conflicts

zach-freyja commented on PlanetNineGateway/iOS/PlanetNineGateway.swift:
```diff
@@ -35,6 +35,35 @@ public class PlanetNineGateway {
+    
+    public func getUser(userUUID: String, gatewayName: String, callback: @escaping (PNUser) -> Void) {
+        let gatewayTimestampTuple = GatewayTimestampTuple(gatewayName: gatewayName)
+        let signature = Crypto().signMessage(message: gatewayTimestampTuple.toString())
```
you should use a guard here so that you don't have to force unwrap on the next line

zach-freyja commented on PlanetNineGateway/iOS/PlanetNineGateway.swift:
```diff
@@ -35,6 +35,35 @@ public class PlanetNineGateway {
+    public func getUser(userUUID: String, gatewayName: String, callback: @escaping (PNUser) -> Void) {
+        let gatewayTimestampTuple = GatewayTimestampTuple(gatewayName: gatewayName)
+        let signature = Crypto().signMessage(message: gatewayTimestampTuple.toString())
+        _ = PlanetNineUser(userUUID: userUUID, gatewayName: gatewayName, timestamp: gatewayTimestampTuple.timestamp, signature: signature!) { pnUser in
```
Looking at this, I think a cleaner way would be to add a class function to `PlanetNineUser` that takes these params and gets the user. That way we can avoid the awkward `_ = ...`

ekcorso replied:
You know, that was present in the code I moved over from the tester app. I'm not 100% sure why the `_ = ... construction` is used, other than maybe to silence a compiler warning. Happy to fix it, but I'd like to do a little research first to make sure I understand.

zach-freyja reviewed (changes requested):
This looks great! Just need to clean up those merge conflicts, and change that force unwrap. Adding the class function to `PlanetNineUser` would be a bonus, but not necessary at this point.

ekcorso committed:
Fix merge conflicts in Network

ekcorso committed:
Add guard statement to getUser

zach-freyja commented on PlanetNineGateway/iOS/Models/Network.swift:
```diff
@@ -123,18 +123,19 @@ class Network: NSObject {
     
     func usePowerAtOneTimeGateway(powerUsageObject: PowerUsage, callback: @escaping (Error?, Data?) -> Void) {
-        let path = "/user/userUUID/\(powerUsageObject.userUUID)/power/gateway/\(powerUsageObject.gatewayName)"
+//        let path = "/user/userUUID/\(powerUsageObject.userUUID)/power/gateway/\(powerUsageObject.gatewayName)"
```
Let's remove this line

zach-freyja commented on PlanetNineGateway/iOS/Models/Network.swift:
```diff
@@ -145,7 +146,8 @@ class Network: NSObject {
     
     func requestTransfer(transferRequestWithSignature: TransferRequestWithSignature, gatewayName: String, callback: @escaping (Error?, Data?) -> Void) {
-        let path = "/user/userUUID/\(transferRequestWithSignature.userUUID)/gateway/\(gatewayName)/transfer/request"
+//        let path = "/user/userUUID/\(transferRequestWithSignature.userUUID)/gateway/\(gatewayName)/transfer/request"
```
Remove this one too

zach-freyja commented on PlanetNineGateway/iOS/Models/Network.swift:
```diff
@@ -188,7 +190,8 @@ class Network: NSObject {
     func approveTransfer(approveTransferWithSignature: ApproveTransferWithSignature, callback: @escaping (Error?, Data?) -> Void) {
         ///user/userId/:userId/transfer/nineumTransactionId/:nineumTransactionId/approve'
-        let path = "/user/userId/\(approveTransferWithSignature.userId)/transfer/nineumTransactionId/\(approveTransferWithSignature.nineumTransactionId)/approve"
+//        let path = "/user/userId/\(approveTransferWithSignature.userId)/transfer/nineumTransactionId/\(approveTransferWithSignature.nineumTransactionId)/approve"
```
and this one

zach-freyja commented on PlanetNineGateway/iOS/Models/UserModel.swift:
```diff
@@ -60,4 +62,8 @@ class UserModel {
     func clearUser() {
 
     }
+    
+    func saveUser(user: PNUser) {
+        Utils().saveEncodableToDefaults(user, key: "user")
+    }
 }
```
We don't want this in the cocoapod.

zach-freyja reviewed (approved):
Looks good Emily! Just a little cleanup, and you're good to merge. Note that I've approved this, so once those commented lines are gone, and the saveUser is gone you can go ahead and merge. Let me know when you do, and we'll publish a new version of the cocoapod.
